src/factoid.rs

```diff
@@ -31,66 +31,98 @@ impl Brain {
     }
 
     pub fn respond(&mut self, input: &str) -> Option<String> {
-        if self.creates_factoid(input) {
-            return self.create_factoid(input);
+        if let Some(factoid) = self.creates_factoid(input) {
+            return self.learn_factoid(factoid);
         }
 
         // Returns None if respond() gave back an empty vec
         self.responders
             .respond(input)
-            .into_iter()
             .choose(&mut self.rng)
     }
 
     pub fn register_responder<T: 'static + Responder>(&mut self, responder: T) {
         self.responders.register_responder(responder)
     }
 
-    fn addressed(&self, input: &str) -> bool {
-        input.starts_with((self.name.to_owned() + ":").as_str())
+    fn addressed<'a>(&self, input: &'a str) -> Option<&'a str> {
```

Comment:
the important part here being that the output `&str` has at most the same lifetime as `input` - without the lifetimes here rustc would infer all three lifetimes to be independent and borrow checking fails